src/sqlfluff/dialects/dialect_bigquery.py

```diff
@@ -365,23 +387,44 @@ class FunctionSegment(BaseSegment):
     """
 
     type = "function"
-    match_grammar = Sequence(
+    match_grammar = OneOf(
         Sequence(
-            AnyNumberOf(
-                Ref("FunctionNameSegment"),
-                max_times=1,
-                min_times=1,
-                exclude=OneOf(
-                    Ref("ValuesClauseSegment"),
+            # Treat fucnctions which take date parts separately
+            # So those functions parse date parts as DatetimeUnitSegment
+            # rather than identifiers.
+            Sequence(
+                Ref("DatePartFunctionNameSegment"),
```

Comment:
Does the parser know (and does it need to know) which parameter(s) of these functions are date parts? I.e. parameter 1, parameter 2, whatever? Or will it accept a date part for *any* parameter of these functions?

Reply:
No it does not need to know.

All this does is say any DatePart function has `"DatePartFunctionNameSegment"` as a higher precedence than the rest of the `"FunctionContentsGrammar"`.

Some functions (e.g. `EXTRACT`) expect it in first place, some (e.g. `DATE_TRUNC`) in second position, and some (e.g. `DATE_DIFF`) in the third position.

The important point is not to exclude any of those, but just to say "if you see something that looks like a datepart, then treat it as such, otherwise carry on and parse it as a `"FunctionContentsGrammar"`.

This is the way ANSI currently handles it.

This does, however, mean `SELECT EXTRACT(year, year) FROM table1` will incorrectly assume both are date parts (when actually it's only the first part and this really should be like `SELECT EXTRACT(year, t.year) FROM table1 t`). We could setup different grammar for each type (1st, 2nd, 3rd) to avoid this but think that adds a lot of complexity for little gain. Arguable in above instance the second `year` should be qualified with a table alias to make it clear.